Add aggregation response

gsfk opened the pull request:
This is an initial proposal for an aggregated response, as discussed in the Beacon Aggregation / Cohorts Scout. 

See [github discussion here](https://github.com/ga4gh-beacon/beacon-v2/discussions/238) for discussion of other options for aggregation. 




### Features:
---
This particular pr offers: 
- a fixed syntax for an aggregate response
- configuration for the response, based on the principle that things you may want to aggregate by (sex, disease, etc) probably already exist as filtering terms. (See the discussion linked above for a proposal without this assumption). 
- changes are non-breaking


Configuration is handled by adding a boolean flag to mark filtering terms as available to be used as fields in a aggregate response (currently the flag is called `aggregatable` although perhaps `isAggregationTerm` is better). Clients can add desired terms to the request in a manner similar to filters:

```json
"aggregationTerms": ["smoking_status", "HP:0030357"]
```

The format of the response is borrowed loosely from the /cohorts response, a minimal example looks like this: 

```json
"aggregateResults": [
	{
		"name": "Smoking status",
		"id": "smoking_status",
		"distribution": {
			"smoker": 123,
			"non-smoker": 333,
			"former smoker": 88
		}
	},
	{
		"name": "Small cell lung carcinoma",
		"id": "HP:0030357",
                "distribution": {
			"HP:0030357": 32  
		}
	}
]
```

- for the resultSet response, there are places to provide aggregate results for the entire response, as well as individually for each resultSet, all are optional. 

- `beaconBooleanResponse` and `beaconCountResponse` do not have a "response" field, so I put this in "responseSummary". I don't find this particularly satisfactory, ideally the information will be in the same place for each response type. So either we (1) pick a uniform location for the top-level aggregation response across all response types, or ideally (2) deprecate these two files and move toward a single response that doesn't vary by granularity. 

- It may happen that an aggregation term applies to a different entity than the one being queried: suppose you query `/biosamples` and ask for a distribution of results by sex. In the response it should be clear which entity is being counted: 
    ```json
    "distribution": {
        "male": 123,  // individuals or biosamples? 
        "female": 456
    }
    ```
    For fields in beacon models this should perhaps already be clear, but this can be made explicit in the `countEntity` field. I'm aware that we're [trying to eliminate "entity" in favour of "entry"](https://github.com/ga4gh-beacon/beacon-v2/issues/23), but: 
    - "entryType" presumably applies only to standard beacon entities (individuals, biosamples... ) but you may want a distribution over other kinds of things (lab tests, medical treatments, imaging types, etc). Currently this is a string only. 
    - `entityType` [still exists in the spec](https://github.com/ga4gh-beacon/beacon-v2/blob/main/framework/src/common/beaconCommonComponents.yaml#L76) although probably this is an oversight 
    - I would welcome a different name that doesn't restrict it to standard beacon entities
    



---
### What's not handled:

- interactions between categories are not handled, i.e. you should **NOT** expect a response like this: 

    ```json
    {
        "smoker && HP:0100526": 12,
        "smoker && !HP:0100526": 34,
        "non-smoker && HP:0100526": 56,
        "non_smoker && !HP:0100526": 78,
    }
    ````

    This kind of response could be returned regardless, but I don't provide any kind of support for it. 

- I didn't include a total count for each category, in part because of possible confusion over what's being counted and what a total means (in a distribution of diseases, is the total the number of disease instances or total individuals? Or something else?) This could be solved by enforcing that total count is total of `countEntity` always, but I was concerned that confusion would still linger, both for users and implementers. 





- I offer no fixed syntax for pointing to json fields in a model (the current state of filtering terms is that this is left to implementers). For related reasons I do not handle filter scope. I invite comments on whether scope in particular is needed or not. 

- There is no attempt to provide a summary response for the variants in the search results. The best format for this is likely to be much different. 

---

This is a draft PR meant to guide discussion in the Aggregation Scout toward a solution, so comments and suggested changes are encouraged, either here or in the discussion linked above.

gsfk committed:
add configurable aggregation response

gsfk committed:
add countEntity to aggregate response

mbaudis commented:
Looks powerful to me. Immediate comment: How about having distributions representing model based mappings, e.g.

```
name: sex
id: individuals.sex.id
distribution:
    NCIT:C16576: 46
    NCIT:C20197: 0
```

... which could either be the default key  ... or an attribute, in line what we'vbe  discussed for filters:

```
name: Small cell lung carcinoma
id: HP:0030357
modelPath: individuals.diseases.id
distribution:
    HP:0030357: 32  
```

I'm a bit confused by the `id: HP:0030357`; this should be either the property or the value consistently? Or real 2 types - simple value counts (times this filter value was observed) or distribution objects for a fixed or ad hoc aggregation.

gsfk commented:
The distribution with one item is a bit weird, I'm willing to admit. We could put some sort of count here instead? Or put in the `totalCount` field that I left out, and don't give a distribution for fields with no actual distribution? Suggestions welcome. My main goal here was a consistent response that would be easy to compare across instances, without the client having to understand anything about what it's looking at. 

The ontology term here is a bit of an edge case, presumably you usually want to see stuff that comes with a distribution of different things (like all diseases, say, not just one of them). But this echoes the ambiguity of the `id` field in filters: for some (ontologies) it is a value to be found in the data, for others (all other filters) it is a field to be searched against. I didn't want to throw away the edge case. 

The main (only?) goal here is to get all the information you could get from repeated count queries, so I wanted to handle the usual filters use cases.

mbaudis commented:
@gsfk Yeah... This again comes from the concept to express 2 different scopes at once - filter id and values - and then their respective counts.

So, IMO, _probably_ the best approach would be just to have the count of matches for the given filter in your (filtered) results, _i.e._ the count you would get when applying this on top of the found data. And something like "biosamples or individuals" is moot since you have the context of the response entity. There are no distributions since the filter id wouldn't be `sex` but `sex-female` (e.g. for a custom filter) or `NCIT:C16576`; or `ageAtDx:<=P18Y` (you'd have to predefine prototypes for numerics etc. but his is in any case necessary ... unless you allow to return the `values: []` here which one could for numerics).

One could add a parameter to specify that `aggregationType: filteringTerm` per term, to allow custom aggregations by the resource - but IMO this is something I would wait for to be requested as a feature; we could just leave it open what the aggregation is based on.

mbaudis commented on framework/src/responses/sections/beaconAggregateResults.yaml:
```diff
@@ -0,0 +1,45 @@
+items:
+  $ref: "#/$defs/AggregateResultsInstance"
+$defs:
+  AggregateResultsInstance:
```
I would change this to `AggregatedResultsInstance` - the `aggregate` has been done ...

mbaudis commented on framework/src/responses/sections/beaconAggregateResults.yaml:
```diff
@@ -0,0 +1,45 @@
+        description: >-
+          Name of this aggregation term
+        type: string
+        example: Genders
```
Only using the _value_ instead of the _field_; e.g. `sex-male` simplifies the structure... But see discussion.

mbaudis commented:
@gsfk ... so the other option is to have model matched parameters w/ their distributions; and as mentioned this is in line with/ the concept of using model paths / JSONpath. But this is again something where it would be _nice_ in theory; but implementers frequently won't follow the model & a "filter match count" is more "behaviorally in line". And we can do this later - important to get a viable version in first.

So: Please re-think if you want to / can re-write the current version for a "bare instance count" option, where the instance is documented as a filter id **but** does not have to be one[^1].

[^1]: We have e.g. genomic aggregations like "variant counts per interval" which could be expressed then in an aggregation response instead of a custom [service response](https://www.progenetix.org/services/intervalFrequencies/?filters=NCIT:C4323) - I would _love_ that.

mbaudis committed:
add aggregation granularity

It seems necessary to add a separate granularity level for aggregate data responses?!

mbaudis commented:
@gsfk What about an additional granularity level? Added here ... https://github.com/ga4gh-beacon/beacon-v2/commit/8b8d15fb3e4c3d0ad4b2ba3a6de272f31e08290f

gsfk commented:
> @gsfk What about an additional granularity level? Added here ... [8b8d15f](https://github.com/ga4gh-beacon/beacon-v2/commit/8b8d15fb3e4c3d0ad4b2ba3a6de272f31e08290f)

I certainly have wondered out loud before where `aggregated` granularity went. But my understanding was that no one wanted it back. There were even explicit comments to this effect on the "use case" document that was circulated last year.

mbaudis commented:
@gsfk Oh well - there was no response format corresponding to it, right? And true, I'll change the spelling `aggregation` -> `aggregated`; rather equal & the other variant has precedence.

mbaudis committed:
aggregation -> aggregated granularity

gsfk commented:
Yes, it was an orphaned part of the spec. `aggregated` granularity matches my own use case, but I can think of others that don't quite fit: 
- someone who wants summary statistics for a full record response
- someone who want summary statistics for a beacon network call with a mix of granularities

So I'm not convinced we need a new response type. The goal here was to add summary stats to the existing responses.

mbaudis committed:
removing CountEntity from beaconAggregateResults

Since this aggregate ... implementation uses standard Beacon requests and responses, counts should always reflect the entry type of the response; i.e. it should be the count of the `individuals` having `biosamples` with a certain histology if using `.../individuals/?...`. While certainly _possible_ it seems pretty confusing to have different entities counted. Also it is not intuitive since the response entity indicates _what_ should be reported on.

mbaudis committed:
Re-designing AggregateResultsInstance

This commit modifies the AggregateResultsInstance to accommodate for different options:

* ontology terms as aggregation terms, where the id corresponds to the value and only the count of its occurrences is relevant
* alphanumeric types, where an id (e.g. age, sex...) can have different values (represented in a `distribution` object)

Also, some prototype definitions for `distribution` are provided:
* a `values` list
* a `distincts` list
* an `items` list where objects can be provided

Examples are given. Other `DistributionItem` properties could be added (e.g. `average` ...). Also, the structure is open for some changes (e.g. just use a generic object instead of the `items` array, defined through `patternProperties` or such).

mbaudis commented:
@gsfk I've adjusted to the view that "people use alphanumeric filters and might want to get more info about the values". Please have a look at the current state of the modified PR, and especially at https://github.com/ga4gh-beacon/beacon-v2/pull/237/commits/54413c4dc84570dd3530ac2bf2ad219e37ff5f48.

mbaudis committed:
Merge pull request #243 from ga4gh-beacon/schema-urgent-fixes

Update from Schema urgent fixes

mbaudis committed:
Aggregate => summary

This commit

* removes the `aggregated` granularity again
* changes much of the wording and naming from "aggragation" to "summary"
    - this leaves "aggregate..." where it still makes sense (i.e. a summary from aggregating over items...)
    - might be incomplete or overdoing it - testing here
* changed "summaryTerms" (now....) to be just of type string, not of necessarily corresponding to filter ids (though those are documented)

jrambla reviewed (changes requested):
We should split the `measures` to `measurements` from this PR into a different PR

mbaudis committed:
Revert "Merge pull request #229 from ga4gh-beacon/schema-urgent-fixes_change_measurement_to_measure"

This reverts commit 559bfe4c652d1b7438871b4b70ffa5319ae9df13, reversing
changes made to d227d242467484e293a182a37df69e2f596ad3db.

mbaudis commented on framework/src/responses/sections/beaconFilteringTermsResults.yaml:
```diff
@@ -75,6 +75,11 @@ $defs:
         items:
           type: string
         minItems: 1
+      aggregatable:
```
We have to think if we should add this here or just have a separate response for anything that can be a "summary" (including where the items are filtering terms) since we (IMO) need additional summary types beyond those existing as filters.

mbaudis commented:
@jrambla True; was one of those tiny-fixes-that-might-break-things & I lost track of the current state per branch. Tried to revert again but maybe @costero-e can help to clean this up :-)

jrambla commented:
I've added some questions about the scope in the discussion #238 
Please respond there.

costero-e committed:
Merge pull request #248 from ga4gh-beacon/develop

Updating main from develop, milestone 2.2.0

mbaudis committed:
trimming aggregation proposal after McGill WS

mbaudis committed:
trimming aggregation proposal after McGill WS

mbaudis committed:
reformatting examples

mbaudis committed:
more example formatting

mbaudis committed:
pre upstream merge conflict resolution

mbaudis committed:
Merge branch 'add-aggregation-response' of https://github.com/ga4gh-beacon/beacon-v2 into add-aggregation-response

mbaudis committed:
Merge pull request #258 from ga4gh-beacon/main

update add-aggregation-response from main

mbaudis committed:
whitespace

no changes but one of those "edited json vs yaml -> json" description changes

mbaudis committed:
(re)adding responseAggregation and resultsAggregation

mbaudis commented:
All: I will close this PR since there have been many discussions & changes of concepts. I'll reopen ASAP a new PR so we can restart the discussion w/ less overhead.
